core/src/main/java/com/linecorp/armeria/common/stream/ByteStreamMessageOutputStream.java

```diff
@@ -0,0 +1,236 @@
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common.stream;
+
+import static java.util.Objects.requireNonNull;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.function.Consumer;
+
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import com.linecorp.armeria.common.HttpData;
+import com.linecorp.armeria.common.annotation.Nullable;
+
+import io.netty.util.concurrent.EventExecutor;
+
+final class ByteStreamMessageOutputStream implements ByteStreamMessage {
+
+    private final StreamMessageAndWriter<HttpData> outputStreamWriter = new DefaultStreamMessage<>();
+    private final ByteStreamMessage delegate = ByteStreamMessage.of(outputStreamWriter);
+
+    private final Consumer<OutputStream> outputStreamConsumer;
+    private final Executor blockingTaskExecutor;
+
+    ByteStreamMessageOutputStream(Consumer<OutputStream> outputStreamConsumer, Executor blockingTaskExecutor) {
+        requireNonNull(outputStreamConsumer, "outputStreamConsumer");
+        requireNonNull(blockingTaskExecutor, "blockingTaskExecutor");
+        this.outputStreamConsumer = outputStreamConsumer;
+        this.blockingTaskExecutor = blockingTaskExecutor;
+    }
+
+    @Override
+    public ByteStreamMessage range(long offset, long length) {
+        delegate.range(offset, length);
+        return this;
+    }
+
+    @Override
+    public boolean isOpen() {
+        return delegate.isOpen();
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return delegate.isEmpty();
+    }
+
+    @Override
+    public long demand() {
+        return delegate.demand();
+    }
+
+    @Override
+    public CompletableFuture<Void> whenComplete() {
+        return delegate.whenComplete();
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super HttpData> subscriber, EventExecutor executor,
+                          SubscriptionOption... options) {
+        final Subscriber<HttpData> outputStreamSubscriber = new OutputStreamSubscriber(
+                subscriber, executor, outputStreamWriter, outputStreamConsumer, blockingTaskExecutor);
+        delegate.subscribe(outputStreamSubscriber, executor, options);
+    }
+
+    @Override
+    public void abort() {
+        delegate.abort();
+    }
+
+    @Override
+    public void abort(Throwable cause) {
+        delegate.abort(cause);
+    }
+
+    private static final class OutputStreamSubscriber implements Subscriber<HttpData>, Subscription {
```

Comment:
Should we remove `Subscription` from the super interfaces? `OutputStreamSubscriber` simply needs to relay the published objects.
```suggestion
    private static final class OutputStreamSubscriber implements Subscriber<HttpData> {
```

Reply:
✅ updated!